src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -740,27 +740,31 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
 
     def fromScala2x = unapplyFn.symbol.exists && (unapplyFn.symbol.owner is Scala2x)
 
-    /** Can `subtp` be made to be a subtype of `tp`, possibly by dropping some
-     *  refinements in `tp`?
+    /** Is `subtp` a subtype of `tp` or of some generalization of `tp`?
+     *  The generalizations of a type T are the smallest set G such that
+     *
+     *   - T is in G
+     *   - If a typeref R in G represents a trait, R's superclass is in G.
+     *   - If a type proxy P is not a reference to a class, P's supertype is in G
      */
     def isSubTypeOfParent(subtp: Type, tp: Type)(implicit ctx: Context): Boolean =
       if (subtp <:< tp) true
       else tp match {
-        case tp: RefinedType => isSubTypeOfParent(subtp, tp.parent)
+        case tp: TypeRef if tp.symbol.isClass =>
+          tp.symbol.is(Trait) && isSubTypeOfParent(subtp, tp.parents.head)
```

Comment:
> I assume tp.symbol.is(Trait) implies tp.symbol.isClass here. If true, it would read better if the guard is changed to tp.symbol.is(Trait) and adapt correspondingly the code.

No that changes the logic. What we need is that if a TypeRef points to a non-trait class, we stop there, so we are not allowed to execute the following case about TypeProxys.